types/index.d.ts

```diff
@@ -1,25 +1,57 @@
-import { RouteOptions, FastifyReply, FastifyRequest } from 'fastify';
+import { RouteOptions, FastifyReply, FastifyRequest, FastifyPluginCallback } from 'fastify';
 
-export declare interface OnRouteEvent extends RouteOptions { }
+type FastifyDiagnosticsChannel = FastifyPluginCallback
 
-export declare interface OnTimeoutEvent {
-  reply: FastifyReply;
-  request: FastifyRequest;
-  connectionTimeout: number;
-}
+declare module 'diagnostics_channel' {
+  type FastifyChannelListener<N, T> = (message: T, name: N) => void;
 
-export declare interface OnErrorEvent {
-  reply: FastifyReply;
-  request: FastifyRequest;
-  error: Error;
-}
+  // @ts-ignore Enforce extending from private Class Channel
+  class FastifyChannel<N extends string, T> extends Channel {
+    subscribe(onMessage: FastifyChannelListener<N, T>): void;
+    unsubscribe(onMessage: FastifyChannelListener<N, T>): void;
+  }
 
-export declare interface OnResponseEvent {
-  reply: FastifyReply;
-  request: FastifyRequest;
+  function channel(name: 'fastify.onRoute'): FastifyChannel<'fastify.onRoute', fastifyDiagnosticsChannel.OnRouteEvent>;
+  function channel(name: 'fastify.onTimeout'): FastifyChannel<'fastify.onTimeout', fastifyDiagnosticsChannel.OnTimeoutEvent>;
+  function channel(name: 'fastify.onError'): FastifyChannel<'fastify.onError', fastifyDiagnosticsChannel.OnErrorEvent>;
+  function channel(name: 'fastify.onResponse'): FastifyChannel<'fastify.onResponse', fastifyDiagnosticsChannel.OnResponseEvent>;
+  function channel(name: 'fastify.onRequest'): FastifyChannel<'fastify.onRequest', fastifyDiagnosticsChannel.OnRequestEvent>;
 }
 
-export declare interface onRequestEvent {
-  reply: FastifyReply;
-  request: FastifyRequest;
+declare namespace fastifyDiagnosticsChannel {
+  export interface OnRouteEvent extends RouteOptions { }
+  
+  export interface OnTimeoutEvent {
+    reply: FastifyReply;
+    request: FastifyRequest;
+    connectionTimeout: number;
+  }
+  
+  export interface OnErrorEvent {
+    reply: FastifyReply;
+    request: FastifyRequest;
+    error: Error;
+  }
+  
+  export interface OnResponseEvent {
+    reply: FastifyReply;
+    request: FastifyRequest;
+  }
+
+  
+  export interface OnRequestEvent {
+    reply: FastifyReply;
+    request: FastifyRequest;
+  }
+  /**
+   * @deprecated Use OnRequestEvent instead
```

Comment:
I'm not sure I understand. You are deprecating and recommending the deprecated usage?

Reply:
`o` vs `O`